### ISSUE TITLE
BIT(n) data type converts to zeros?

### ISSUE THREAD

**Issue by Paril** [sidorares/node-mysql2#548]
So I'm having an odd issue related to the BIT data type, and its behavior is acting quite strange compared to the npm mysql driver, which I used prior to mysql2.

In that version, their prepared statements allowed me to interact with the BIT data type as if it was a number with no issue - it returned me an integer from the data base, and inserting/using bitwise operations worked just fine.

For some reason, when I perform the same queries with mysql2, the numbers in the prepared statements are being converted to zero (as expected)... but the values actually inserted into the database are 1 and 0110000, (for BIT(1) and BIT(7)) respectively, even though the MySQL event log + query log on my end show that it should be zero.

These are the two arguments passed to pool.execute, which I log in my own file:

```{"query":"INSERT INTO `chat_messages` (sender, recipient, text, time, type, flags) VALUES (?, ?, ?, ?, ?, ?)","arguments":[139580,1,"Hello from bot",1492249595,0,0]}```

And this is what the MySQL General Log logged:

```		  413 Prepare	INSERT INTO `chat_messages` (sender, recipient, text, time, type, flags) VALUES (?, ?, ?, ?, ?, ?)
		  413 Execute	INSERT INTO `chat_messages` (sender, recipient, text, time, type, flags) VALUES ('139580', '1', 'Hello from bot', '1492249595', '0', '0')```

Looks all good, I expect zero and zero to be inserted - however...

![](http://i.imgur.com/nPqFp1O.png)

The issue appears to be in the way the prepared statements function. Instead of allowing numbers to be numbers, all arguments are forced to be strings, which confuses the MySQL parser since it is interpreting '0' as a bitset instead of the number zero. If I copy the same query and perform it without the quotes around the numbers, it works a treat.

Is there any way to disable the quotes around non-string arguments?

**Comment by Paril** [comment]
I've discovered that I can temporarily work around this by doing this:

```INSERT INTO `chat_messages` (sender, recipient, text, time, type, flags) VALUES (?, ?, ?, ?, CAST(? AS UNSIGNED), CAST(? AS UNSIGNED))```

It's a bit ugly, but to be fair I did use this method below in another query:

```SELECT `id`, `sender`, `recipient`, CAST(`type` AS UNSIGNED) AS `type`, CAST(`flags` AS UNSIGNED) AS `flags` FROM `chat_messages` WHERE id = ?```

I'm just wondering if there's anything that can be done about prepared statements converting the arguments to string; it seems unnecessary, prepared statements can support integral arguments can't they? I use a similar thing on PHP via PDO, and it seems to work fine with bitsets/integral values.

**Comment by sidorares** [comment]
yes, I would like `.execute()` to serialise numbers as numbers when possible

related discussions: #353 and #446
